---
title: "Paired-Tag QC"
author: ""
date: ""
output: html_document
---
<style type="text/css">
.main-container {
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r setup, include=FALSE}
library(knitr)
library(varhandle)
knitr::opts_chunk$set(echo = TRUE)
input<-read.csv("QC_library_list.xls", sep="\t", head=F)
```
### Project summary

File annotations are in "QC_Library_List.xls" file and output a summary file "QC_Summary_metaData_Table.xls".

```{r read_information, echo=FALSE, fig.width=18, fig.height=9}
colnames(input)<-c("DNA_ID","RNA_ID","Genome","Annotations")
n_libs=dim(input)[1]
file_name=paste("QC_Summary_", as.vector(input[1,1]), "_", as.vector(input[1,2]), ".xls", sep="")
meta<-read.csv(file_name, sep="\t", head=F)
rownames(meta)<-c("Total_Reads_DNA", "Total_Reads_RNA", "Valid_Reads_DNA", "Valid_Reads_RNA", "Frac_Valid_Reads_DNA", "Frac_Valid_Reads_RNA","Assigned_Reads_DNA", "Assigned_Reads_RNA", "Frac_Assigned_Reads_DNA", "Frac_Assigned_Reads_RNA", "Mapped_Reads_DNA", "Mapped_Reads_RNA", "Mapping_Rate_DNA", "Mapping_Rate_RNA","Unique_Reads_DNA", "Unique_Reads_RNA", "Dup_Rates_DNA", "Dup_Reads_RNA", "PF_Cells_DNA", "PF_Cells_RNA", "PF_Cells_Paired", "RNA_UMI_Cutoff", "DNA_Frag_Cutoff", "Med_UMI_RNA", "Med_Genes_RNA", "Med_Frag_DNA", "Med_Features_DNA")
colnames(meta)<-c("N", paste("[",input[1,1],"_", input[1,2], "](", "Per_lib_QC_",input[1,1],"_", input[1,2], ".html)",sep=""))

if(n_libs>1){
  for(i in c(2:n_libs)){
    print(i)
    file_name=paste("QC_Summary_", as.vector(input[i,1]), "_", as.vector(input[i,2]), ".xls", sep="")
    meta.add<-read.csv(file_name, sep="\t", head=F)
    meta<-cbind(meta, as.vector(meta.add[,2]))
    colnames(meta)[1+i]<-paste("[",input[i,1],"_", input[i,2], "](", "Per_lib_QC_",input[i,1],"_", input[i,2], ".html)",sep="")
  }
}
write.table(meta[,2:dim(meta)[2]], col.names=T, row.names=T, sep="\t", quote=F, file="QC_Summary_metaData_Table.xls" )
```


```{r plot, include=TRUE,echo=FALSE, fig.width=12, fig.height=8}
meta.s<-meta[,2:(n_libs+1)]
#print(head(meta.s))
par(mfrow=c(3,4))
# frac valid DNA
frac<-as.numeric(unfactor(meta.s[3,]))/as.numeric(unfactor(meta.s[1,]))*100
col<-rep("firebrick3", n_libs)
col[frac>60]<-"goldenrod3";col[frac>70]<-"darkolivegreen4";col[frac>80]<-"forestgreen";col[frac>90]<-"deepskyblue3"
names(frac)<-as.vector(input$V1)
barplot(frac, ylim=c(0,100), ylab="Valid Reads (%) - DNA", border=F, col=col, main="Fraction of Valid Reads DNA", las=2, names.arg=as.vector(input$DNA_ID))
## frac assigned DNA
frac<-as.numeric(unfactor(meta.s[7,]))/as.numeric(unfactor(meta.s[3,]))*100
col<-rep("firebrick3", n_libs)
col[frac>60]<-"goldenrod3";col[frac>70]<-"darkolivegreen4";col[frac>80]<-"forestgreen";col[frac>90]<-"deepskyblue3"
barplot(frac, ylim=c(0,100), ylab="Assigned Reads (%) - DNA", border=NA, col=col, main="Fraction of Assigned Reads DNA", las=2, names.arg=as.vector(input$DNA_ID))
## frac mapped DNA
frac<-as.numeric(unfactor(meta.s[11,]))/as.numeric(unfactor(meta.s[7,]))*100
col<-rep("firebrick3", n_libs)
col[frac>20]<-"goldenrod3";col[frac>40]<-"darkolivegreen4";col[frac>65]<-"forestgreen";col[frac>80]<-"deepskyblue3"
barplot(frac, ylim=c(0,100), ylab="Mapped Reads (%) - DNA", border=NA, col=col, main="Fraction of Mapped Reads DNA", las=2, names.arg=as.vector(input$DNA_ID))
## frac dup DNA
frac<-100-as.numeric(unfactor(meta.s[15,]))/as.numeric(unfactor(meta.s[11,]))*100
col<-rep("firebrick3", n_libs)
col[frac<70]<-"goldenrod3";col[frac<50]<-"darkolivegreen4";col[frac<30]<-"forestgreen";col[frac<10]<-"deepskyblue3"
barplot(frac, ylim=c(0,100), ylab="Duplication Rates (%) - DNA", border=NA, col=col, main="Duplication Rates DNA", las=2, names.arg=as.vector(input$DNA_ID))

## frac valid DNA
frac<-as.numeric(unfactor(meta.s[4,]))/as.numeric(unfactor(meta.s[2,]))*100
col<-rep("firebrick3", n_libs)
col[frac>60]<-"goldenrod3";col[frac>70]<-"darkolivegreen4";col[frac>80]<-"forestgreen";col[frac>90]<-"deepskyblue3"
names(frac)<-as.vector(input$V1)
barplot(frac, ylim=c(0,100), ylab="Valid Reads (%) - RNA", border=F, col=col, main="Fraction of Valid Reads RNA", las=2, names.arg=as.vector(input$RNA_ID))
## frac assigned RNA
frac<-as.numeric(unfactor(meta.s[8,]))/as.numeric(unfactor(meta.s[4,]))*100
col<-rep("firebrick3", n_libs)
col[frac>60]<-"goldenrod3";col[frac>70]<-"darkolivegreen4";col[frac>80]<-"forestgreen";col[frac>90]<-"deepskyblue3"
barplot(frac, ylim=c(0,100), ylab="Assigned Reads (%) - RNA", border=NA, col=col, main="Fraction of Assigned Reads RNA", las=2, names.arg=as.vector(input$RNA_ID))
## frac mapped RNA
frac<-as.numeric(unfactor(meta.s[12,]))/as.numeric(unfactor(meta.s[8,]))*100
col<-rep("firebrick3", n_libs)
col[frac>20]<-"goldenrod3";col[frac>40]<-"darkolivegreen4";col[frac>65]<-"forestgreen";col[frac>80]<-"deepskyblue3"
barplot(frac, ylim=c(0,100), ylab="Mapped Reads (%) - RNA", border=NA, col=col, main="Fraction of Mapped Reads RNA", las=2, names.arg=as.vector(input$RNA_ID))
## frac dup RNA
frac<-100-as.numeric(unfactor(meta.s[16,]))/as.numeric(unfactor(meta.s[12,]))*100
col<-rep("firebrick3", n_libs)
col[frac<70]<-"goldenrod3";col[frac<50]<-"darkolivegreen4";col[frac<30]<-"forestgreen";col[frac<10]<-"deepskyblue3"
barplot(frac, ylim=c(0,100), ylab="Duplication Rates (%) - RNA", border=NA, col=col, main="Duplication Rates RNA", las=2, names.arg=as.vector(input$RNA_ID))

# n_UMI_RNA
num<-as.numeric(unfactor(meta.s[24,]))
barplot(num, ylim=c(1, max(num)*5), ylab="# of UMI per nuclei (RNA)", border=NA, col="grey", main="# of UMI (RNA)", las=2, names.arg=as.vector(input$RNA_ID), log="y")
# n_genes_RNA
num<-as.numeric(unfactor(meta.s[25,]))
barplot(num, ylim=c(1, max(num)*5), ylab="# of genes per nuclei (RNA)", border=NA, col="grey", main="# of genes (RNA)", las=2, names.arg=as.vector(input$RNA_ID), log="y")
# n_frag_DNA
num<-as.numeric(unfactor(meta.s[26,]))
barplot(num, ylim=c(1, max(num)*5), ylab="# of fragments per nuclei (DNA)", border=NA, col="grey", main="# of fragments (DNA)", las=2, names.arg=as.vector(input$DNA_ID), log="y")
# n_features_DNA
num<-as.numeric(unfactor(meta.s[27,]))
barplot(num, ylim=c(1, max(num)*5), ylab="# of features per nuclei (DNA)", border=NA, col="grey", main="# of features (DNA)", las=2, names.arg=as.vector(input$DNA_ID), log="y")

```

\
This table summary the sub-libraries processed under this folder. \
```{r table, echo=FALSE}
kable(meta[2:dim(meta)[2]])

```

